kitsune/search/views.py

```diff
@@ -54,10 +54,113 @@ def jsonp_is_valid(func):
+
+    Further, this verifies the ``JSON_CALLBACK`` if there is one and if it's not
+    valid, it returns an error response.
+
+    """
+    def _markup_json(request, *args, **kwargs):
+        request.IS_JSON = request.GET.get('format') == 'json'
+        if request.IS_JSON:
+            request.JSON_CALLBACK = request.GET.get('callback', '').strip()
+        else:
+            request.JSON_CALLBACK = ''
+        if request.IS_JSON:
+            request.CONTENT_TYPE = (
+                'application/x-javascript' if request.JSON_CALLBACK else 'application/json')
+        else:
+            request.CONTENT_TYPE = 'text/html'
+
+        # Check callback is valid
+        if request.JSON_CALLBACK and not jsonp_is_valid(request.JSON_CALLBACK):
+            return HttpResponse(
+                json.dumps({'error': _('Invalid callback function.')}),
+                content_type=request.CONTENT_TYPE,
+                status=400)
+
+        return fun(request, *args, **kwargs)
+    return _markup_json
+
 
+def _es_down_template(request, *args, **kwargs):
+    """Returns the appropriate "Elasticsearch is down!" template"""
+    return 'search/mobile/down.html' if request.MOBILE else 'search/down.html'
 
+
+def build_results_list(pages, is_json):
+    """Takes a paginated search and returns results List
+
+    Handles wiki documents, questions and contributor forum posts.
+
+    :arg pages: paginated S
+    :arg is_json: whether or not this is generated results for json output
+
+    :returns: list of dicts
+
+    """
+    results = []
+    for rank, doc in enumerate(pages, pages.start_index()):
+        if doc['model'] == 'wiki_document':
+            summary = _build_es_excerpt(doc)
+            if not summary:
+                summary = doc['document_summary']
+            result = {
+                'title': doc['document_title'],
+                'type': 'document'}
+
+        elif doc['model'] == 'questions_question':
+            summary = _build_es_excerpt(doc)
+            if not summary:
+                # We're excerpting only question_content, so if the query matched
+                # question_title or question_answer_content, then there won't be any
+                # question_content excerpts. In that case, just show the question--but
+                # only the first 500 characters.
+                summary = bleach.clean(doc['question_content'], strip=True)[:500]
+
+            result = {
+                'title': doc['question_title'],
+                'type': 'question',
+                'is_solved': doc['question_is_solved'],
+                'num_answers': doc['question_num_answers'],
+                'num_votes': doc['question_num_votes'],
+                'num_votes_past_week': doc['question_num_votes_past_week']}
+
+        else:
+            summary = _build_es_excerpt(doc, first_only=True)
+            result = {
+                'title': doc['post_title'],
+                'type': 'thread'}
```

Comment:
Is this `else` clause a catch all? Or is it for a specific document type? If it is a catchall, this doesn't seem robust enough. If it is for a specific type, I think it should be an `elif doc['model'] == 'that':`' instead. Maybe even have logging or an exception for unknown models as a sort of sanity assertion?